web/dev_server_client.js

```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+
+// TODO(yjbanov): this should be moved served by the dev server and not included in the application code.
+
+function printStats() {
+    console.log(document.querySelectorAll('*').length, 'elements');
+}
+
+class ButterflyModule {
+    constructor(moduleName, hostElement) {
+      this.moduleName = moduleName;
+      this.hostElement = hostElement;
+    }
+
+    run() {
+        let eventTypes = ["click", "keyup"];
+        eventTypes.forEach((type) => {
+            this.hostElement.addEventListener(type, (event) => {
+                this._handleEvent(type, event);
+            });
+        });
+        this._initialize();
+        this._renderFrame();
+    }
+
+    /// Synchronously (i.e. blocks the UI thread) invokes a platform channel [method] and returns the result.
+    ///
+    /// [args] are encoded as JSON. The result is decoded as JSON.
+    _invokePlatformChannelMethod(method, args) {
+      let xhr = new XMLHttpRequest();
+      xhr.open('POST', `/__butterfly_dev_channel__/${this.moduleName}/${method}`, false /* synchronous */);
+      xhr.setRequestHeader("Content-Type", "text/plain;charset=UTF-8");
+      xhr.send(JSON.stringify(args));
+      let result = JSON.parse(xhr.responseText);
+      if (result != null && result.hasOwnProperty('error')) {
+        throw result['error'];
+      }
+      return result;
+    }
+
+    _initialize() {
+        let diff = this._invokePlatformChannelMethod('initialize', '');
+    }
+
+    _renderFrame() {
+        let diff = this._invokePlatformChannelMethod('render-frame', '');
+        if (!diff) {
+            return;
+        }
+        let jsonParseEnd = performance.now();
+
+        if (diff.hasOwnProperty("create")) {
+            let createStart = performance.now();
+            this.hostElement.innerHTML = diff["create"];
+            let createEnd = performance.now();
+            printPerf('create', createStart, createEnd);
+        } else if (diff.hasOwnProperty("update")) {
+            let updateStart = performance.now();
+            this._applyElementUpdate(this.hostElement.firstElementChild, diff["update"]);
+            let updateEnd = performance.now();
+            printPerf('update', updateStart, updateEnd);
+        }
+        console.timeStamp('End apply diff');
+    }
+
+    _handleEvent(type, event) {
+        // Look for the nearest parent with a _bid, then dispatch to it.
+        let baristaId = null;
+        let parent = event.target;
+        while(baristaId == null && parent && parent != document) {
+            baristaId = parent.getAttribute("_bid");
+            parent = parent.parentNode;
+        }
+        if (baristaId) {
+            this._invokePlatformChannelMethod('dispatch-event', serializeEvent(type, baristaId, event));
+            this._renderFrame();
+        } else {
+            console.log(">>> caught event on target with no _bid:", event.target);
```

Comment:
I'm not sure this will always be an error. Need to think about this. It depends on whether we keep using the global handler trick or switch to per-element handlers.